addons/bus/static/src/services/worker_service.js

```diff
@@ -0,0 +1,133 @@
+import { browser } from "@web/core/browser/browser";
+import { registry } from "@web/core/registry";
+import { Deferred } from "@web/core/utils/concurrency";
+import { session } from "@web/session";
+import { isIosApp } from "@web/core/browser/feature_detection";
+
+export const WORKER_STATE = Object.freeze({
+    UNINITIALIZED: "UNINITIALIZED",
+    INITIALIZING: "INITIALIZING",
+    INITIALIZED: "INITIALIZED",
+    FAILED: "FAILED",
+});
+
+export const workerService = {
+    dependencies: ["bus.parameters"],
+
+    start(env, { "bus.parameters": params }) {
+        let worker;
+        let isUsingSharedWorker = browser.SharedWorker && !isIosApp();
+        let state = WORKER_STATE.UNINITIALIZED;
+        let connectionInitializedDeferred = null;
+
+        function startWorker() {
+            if (!connectionInitializedDeferred) {
+                connectionInitializedDeferred = new Deferred();
+            }
+            state = WORKER_STATE.INITIALIZING;
+            let workerURL = `${params.serverURL}/bus/websocket_worker_bundle?v=${session.websocket_worker_version}`;
+            if (params.serverURL !== window.origin) {
+                // Worker service can be loaded from a different origin than the
+                // bundle URL. The Worker expects an URL from this origin, give
+                // it a base64 URL that will then load the bundle via "importScripts"
+                // which allows cross origin.
+                const source = `importScripts("${workerURL}");`;
+                workerURL = "data:application/javascript;base64," + window.btoa(source);
+            }
+            const workerClass = isUsingSharedWorker ? browser.SharedWorker : browser.Worker;
+            worker = new workerClass(workerURL, {
+                name: isUsingSharedWorker ? "odoo:bus_shared_worker" : "odoo:bus_worker",
+            });
+            worker.onerror = onInitError;
+            _registerHandler(handleMessage);
+            if (isUsingSharedWorker) {
+                worker.port.start();
+            }
+            _send("BASE:INIT");
+            // if no reply, treat as there's an error
+            // setTimeout(onInitError, 5000);
```

Comment:
Do we want to keep it or not? Also, isn’t it redundant with the `onerror` handler? I’d just keep the "handshake" and remove the error handler (but the `onInitError` should still be called if we don't receive the expected event from the worker).

I wonder if we need the `BASE:INIT` event at all. Why not just send `WORKER:ALIVE` to each newly connected client? This way, we can get rid of the `BaseWorker`.